File: src/arduinobot_remote/arduinobot_remote/task_server.py
#! /usr/bin/env python3
import rclpy
from rclpy.node import Node
from rclpy.action import ActionServer
from arduinobot_msgs.action import ArduinobotTask
from moveit.planning import MoveItPy
from moveit.core.robot_state import RobotState
import numpy as np


class TaskServer(Node):
    def __init__(self):
        super().__init__("task_server")
        self.actionServer = ActionServer(
            self,
            ArduinobotTask,
            "task_server",
            self.goalCallback,
        )
        self.arduinobot = MoveItPy(node_name="moveit_py")
        self.arduinobot_arm = self.arduinobot.get_planning_component("arm")
        self.arduinobot_gripper = self.arduinobot.get_planning_component("gripper")

        self.get_logger().info("Starter task_server")

    def goalCallback(self, goalHandle):
        self.get_logger().info(
            f"New Goal received Task#: {goalHandle.request.task_number}"
        )

        result = False

        arm_state = RobotState(self.arduinobot.get_robot_model())
        gripper_state = RobotState(self.arduinobot.get_robot_model())

        arm_joint_goal = []
        gripper_joint_goal = []

        if goalHandle.request.task_number == 0:
            arm_joint_goal = np.array([0.0, 0.0, 0.0])
            gripper_joint_goal = np.array([-0.7, 0.7])
        elif goalHandle.request.task_number == 1:
            arm_joint_goal = np.array([-1.14, -0.6, -0.07])
            gripper_joint_goal = np.array([0.0, 0.0])
        elif goalHandle.request.task_number == 2:
            arm_joint_goal = np.array([-1.57, 0.0, -0.9])
            gripper_joint_goal = np.array([0.0, 0.0])
        else:
            self.get_logger().error("Unknown Task")
            result.success = result
            return result

        arm_state.set_joint_group_positions("arm", arm_joint_goal)
        gripper_state.set_joint_group_positions("gripper", gripper_joint_goal)

        self.arduinobot_arm.set_start_state_to_current_state()
        self.arduinobot_gripper.set_start_state_to_current_state()

        self.arduinobot_arm.set_goal_state(robot_state=arm_state)
        self.arduinobot_gripper.set_goal_state(robot_state=gripper_state)

        arm_plan = self.arduinobot_arm.plan()
        gripper_plan = self.arduinobot_gripper.plan()

        if arm_plan or gripper_plan:
            result = True

        if arm_plan:
            self.arduinobot.execute(arm_plan.trajectory, controllers=[])
        else:
            self.get_logger().error("Unable to plan arm movement")

        if gripper_plan:
            self.arduinobot.execute(gripper_plan.trajectory, controllers=[])
        else:
            self.get_logger().error("Unable to plan gripper movement")

        goalHandle.succeed()
        result = ArduinobotTask.Result()
        result.success = result
        return result

    def cancelCallback(self, goalHandle):
        pass


def main():
    rclpy.init()
    taskServer = TaskServer()
    rclpy.spin(taskServer)
    TaskServer.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/arduinobot_remote/src/task_server.cpp
#include <rclcpp/rclcpp.hpp>
#include <rclcpp_action/rclcpp_action.hpp>
#include <rclcpp_components/register_node_macro.hpp>
#include <moveit/move_group_interface/move_group_interface.hpp>
#include <arduinobot_msgs/action/arduinobot_task.hpp>
#include <memory>
#include <thread>

using namespace std::placeholders;

namespace arduinobot_cpp_examples
{
    class TaskServer : public rclcpp::Node
    {
    public:
        explicit TaskServer(const rclcpp::NodeOptions &options = rclcpp::NodeOptions())
            : Node("simple_action_server", options)
        {
            RCLCPP_INFO(get_logger(), "Starting Arduinotbot Task action cpp server");

            action_server_ = rclcpp_action::create_server<arduinobot_msgs::action::ArduinobotTask>(
                this,
                "task_server",
                std::bind(&TaskServer::goalCallback, this, _1, _2),
                std::bind(&TaskServer::cancelCallback, this, _1),
                std::bind(&TaskServer::acceptedCallback, this, _1));
        }

    private:
        rclcpp_action::Server<arduinobot_msgs::action::ArduinobotTask>::SharedPtr action_server_;
        std::shared_ptr<moveit::planning_interface::MoveGroupInterface> arm_move_group_, gripper_move_group_;

        rclcpp_action::GoalResponse goalCallback(const rclcpp_action::GoalUUID &uuid,
                                                 std::shared_ptr<const arduinobot_msgs::action::ArduinobotTask::Goal> goal)
        {
            RCLCPP_INFO_STREAM(get_logger(), "Goal request - task_number: " << goal->task_number);
            (void)uuid;
            if(goal->task_number > 2)
                return rclcpp_action::GoalResponse::REJECT;
            return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
        }

        rclcpp_action::CancelResponse cancelCallback(const std::shared_ptr<rclcpp_action::ServerGoalHandle<arduinobot_msgs::action::ArduinobotTask>> goalHandle)
        {
            RCLCPP_INFO(get_logger(), "Cancellation request recived");
            if (arm_move_group_)
            {
                arm_move_group_->stop();
            }
            if (gripper_move_group_)
            {
                gripper_move_group_->stop();
            }
            (void)goalHandle;
            return rclcpp_action::CancelResponse::ACCEPT;
        }

        void acceptedCallback(const std::shared_ptr<rclcpp_action::ServerGoalHandle<arduinobot_msgs::action::ArduinobotTask>> goalHandle)
        {
            std::thread(std::bind(&TaskServer::execute, this, _1), goalHandle).detach();
        }

        void execute(const std::shared_ptr<rclcpp_action::ServerGoalHandle<arduinobot_msgs::action::ArduinobotTask>> goal_handle)
        {
            RCLCPP_INFO(get_logger(), "Executing goal");
            if (!arm_move_group_)
            {
                arm_move_group_ = std::make_shared<moveit::planning_interface::MoveGroupInterface>(shared_from_this(), "arm");
            }
            if (!gripper_move_group_)
            {
                gripper_move_group_ = std::make_shared<moveit::planning_interface::MoveGroupInterface>(shared_from_this(), "gripper");
            }
            arm_move_group_->setStartState(*arm_move_group_->getCurrentState());
            gripper_move_group_->setStartState(*gripper_move_group_->getCurrentState());

            auto result = std::make_shared<arduinobot_msgs::action::ArduinobotTask::Result>();
            auto &success = result->success;

            const auto goal = goal_handle->get_goal();
            auto feedback = std::make_shared<arduinobot_msgs::action::ArduinobotTask::Feedback>();

            std::vector<double> arm_joint_goal{};
            std::vector<double> gripper_joint_goal{};

            switch (goal->task_number)
            {
            case 0:
                arm_joint_goal = {0.0, 0.0, 0.0};
                gripper_joint_goal = {-0.7, 0.7};
                break;
            case 1:
                arm_joint_goal = {-1.14, -0.6, -0.07};
                gripper_joint_goal = {0.0, 0.0};
                break;
            case 2:
                arm_joint_goal = {-1.57, 0.0, -0.9};
                gripper_joint_goal = {0.0, 0.0};
                break;
            default:
                RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "AUnknown Task Number");
                goal_handle->succeed(result);
                return;
            }            
            if (!arm_move_group_->setJointValueTarget(arm_joint_goal))
            {
                RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Arm position unreachable");
            }
            else
            {
                success = true;
                moveit::planning_interface::MoveGroupInterface::Plan arm_plan;
                if (arm_move_group_->plan(arm_plan) == moveit::core::MoveItErrorCode::SUCCESS)
                {
                    arm_move_group_->move();
                }
                else
                {
                    RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Arm movement planning failed");
                }
            }

            if (!gripper_move_group_->setJointValueTarget(gripper_joint_goal))
            {
                RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "gripper position unreachable");
            }
            else
            {
                success = true;
                moveit::planning_interface::MoveGroupInterface::Plan gripper_plan;
                if (gripper_move_group_->plan(gripper_plan) == moveit::core::MoveItErrorCode::SUCCESS)
                {
                    gripper_move_group_->move();
                }
                else
                {
                    RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Gripper movement planning failed");
                }
            }
            
            if (rclcpp::ok())
            {
                goal_handle->succeed(result);
                RCLCPP_INFO(get_logger(), "Goal succeeded");
            }
        }
    };

}

RCLCPP_COMPONENTS_REGISTER_NODE(arduinobot_cpp_examples::TaskServer)
